sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sns/SnsIO.java

```diff
@@ -62,310 +58,272 @@
-                || !spec.getRetryConfiguration().getRetryPredicate().test(ex)) {
-              SNS_WRITE_FAILURES.inc();
-              LOG.info("Unable to publish message {} due to {} ", request.message(), ex);
-              throw new IOException("Error writing to SNS (no attempt made to retry)", ex);
-            }
+      public void processElement(ProcessContext context, BoundedWindow elementWindow)
+          throws IOException {
+        T element = context.element();
+        Instant timestamp = context.timestamp();
+        PaneInfo paneInfo = context.pane();
+
+        WindowedValue<T> windowedValue =
+            WindowedValue.of(element, timestamp, elementWindow, paneInfo);
+
+        PublishRequest request = spec.getPublishRequestFn().apply(element);
+        client.publish(request).whenComplete(whenCompleteAction(windowedValue));
 
-            if (!BackOffUtils.next(sleeper, backoff)) {
-              throw new IOException(
-                  String.format(
-                      "Error writing to SNS after %d attempt(s). No more attempts allowed",
-                      attempt),
-                  ex);
-            } else {
-              // Note: this used in test cases to verify behavior
-              LOG.warn(String.format(RETRY_ATTEMPT_LOG, attempt), ex);
+        checkForFailures();
+      }
+
+      private SerializableBiConsumer<PublishResponse, ? super Throwable> whenCompleteAction(
+          final WindowedValue<T> windowedValue) {
+
+        T element = windowedValue.getValue();
+
+        return (response, exception) -> {
+          if (spec.getThrowOnError()) {
+            if (exception != null) {
+              failures.offer(
+                  new SnsWriteException(
+                      String.format("Error while publishing message: %s", element), exception));
+            } else if (!response.sdkHttpResponse().isSuccessful()) {
+              String customMessage =
+                  response.sdkHttpResponse().statusText().isPresent()
+                      ? String.format(
+                          "Reason for failure: %s", response.sdkHttpResponse().statusText().get())
+                      : "";
+
+              failures.offer(
+                  new SnsWriteException(
+                      String.join(
+                          ERROR_MESSAGE_DELIMITER,
+                          String.format("Failure in publishing message: %s.", element),
+                          customMessage)));
             }
           }
+
+          WrappedSnsResponse<T> wrappedSnsResponse =
+              exception != null
+                  ? WrappedSnsResponse.ofError(element, exception)
+                  : WrappedSnsResponse.of(element, response);
+
+          results.offer(KV.of(wrappedSnsResponse, windowedValue));
+        };
+      }
+
+      @FinishBundle
+      public void finishBundle(FinishBundleContext c) throws IOException {
+        checkForFailures();
+
+        int size = results.size();
```

Comment:
Is it possible that the size of `results` will be less than number of sent messages in current bundle?